### ISSUE TITLE
Using String.format()

### ISSUE THREAD

**Issue by Nixx162** [nus-cs2103-AY2324S1/forum#147]
When using String.format(), it is natural to use magic literals. For example: 

`String.format("You used \"%s\" as an argument. A bye command shouldn't have any arguments", desc)`

Is this an acceptable piece of code or is String.format() is not a good method to use?

**Comment by damithc** [comment]
@Nixx162 are you asking if _`String.format` should be avoided because it forces us to use magic strings_?
If so, note that `String.format` does NOT force us to use magic strings. e.g., `String.format(MSG_INCORRECT_ARGUMENT, desc)` works as well.

Separately, we can discuss whether to use OK to use a magic string with `String.format`. Others are welcome to pitch in with your opinions on on that.

**Comment by Nixx162** [comment]
Ah yes, I believe that using constants on String.format() will really only obscure readability. I suppose the real question we should be asking is: how far is too far when enforcing the no magic literals rule?

**Comment by damithc** [comment]
> Ah yes, I believe that using constants on String.format() will really only obscure readability. I suppose the real question we should be asking is: how far is too far when enforcing the no magic literals rule?

Yup. It's not part of our coding standard, which means it's not really a _rule_. Instead, it is a guideline for improving code quality, which you can choose to follow on a case-by-case basis.

**Comment by woojiahao** [comment]
I think this really boils down to what is defined as a magic literal. An interesting paper I found was this: https://www.sciencedirect.com/science/article/pii/S0950584922000908

And it defines magic literals as 

> Magic literals are such values used without an explicit explanation of their meaning.

So the key component to what makes magic literals magic is the fact that there is a lack of **explicit** explanation of their meaning. In the case of `String.format()`, the strings that you use are self-explanatory, they are simply the contents to print and the interpolated variables should be explained in their own ways. So I don't think that this is a magic literal and it does not affect the readability of the code.

As for whether it is alright to use magic strings, I see this as being a similar problem to writing SQL code in another language. For instance, is it better to include the SQL query directly like (using Go to illustrate but the idea applies):

```go
db.Query(`
SELECT *
FROM users
WHERE username = $1
`, username)
```

Or if it's better to abstract this query into a separate string such as:

```go
const QUERY = `
SELECT *
FROM users
WHERE username = $1
`

db.Query(QUERY, username)
```

I think in this case, having an extra layer of abstraction might actually make it harder to understand the intent of the code and make it harder to trace. You could imagine the `QUERY` being stored in a separate file so you will have to jump between files to figure out these magic literals. I think the SQL string is self-explanatory, albeit slightly long but it doesn't really violate SLAP since it's still along the same level of abstraction.

This is just my take on this though, would love to hear what others think!

**Comment by SungMatt** [comment]
The primary motivation behind eliminating these magic literals is to ensure that future developers can seamlessly continue your work. The goal is to enhance readability, and as long as we uphold this principle, I'm confident your code will be in excellent shape.

**Comment by damithc** [comment]
To add, although may not be a strict 'magic' literal situation, another case where a NAMED_CONSTANT may be worthy of consideration is when the same literal is used multiple times. Moving it to one location as a NAMED_CONSTANT can make future updates to it easier and less error-prone.

**Comment by woojiahao** [comment]
> Moving it to one location as a NAMED_CONSTANT can make future updates to it easier and less error-prone.

I agree, however, I might argue that if it's used in `String.format()`, it's unlikely to be reused so moving it to another location could be considered as a premature optimisation that might not be necessary. I think this really boils down to the context and use case of the literal.

**Comment by damithc** [comment]
> so moving it to another location could be considered as a premature optimisation that might not be necessary.

That's right @woojiahao. I didn't mean _in anticipation of_ reuse; one should wait till it is _actually being reused_ and then decide on based on case-by-case.

**Comment by IsabelChong** [comment]
Hi @Nixx162, just a senior from the previous batch. I see that this has been dormant for quite some time and there is great discussion above. Do close the issue if it has been resolved, or add on for any more related questions <:
